### PR TITLE
Better visualisation interface

### PR DESCRIPTION
**Pull request by alecokas**
* Better naming of the CLI arguments
* More modular and versatile image creation
*  Better Docstrings and general clean-up
* Fixed `snapshot_reconstruction`

### CODE REVIEW

**Comment by alecokas** [comment]
Whoever reviews this, please check that it won't break training

**Comment by alecokas** [comment]
Right now the images are all saved into the main directory - I'll address this soon in my next PR

File: main_viz.py [hunk 1]
```diff
@@ -45,28 +39,56 @@ def read_capacity_from_file(path_to_specs):
...
+    else:
+        num_samples = 1
+    data_loader = get_dataloaders(dataset, shuffle=False, batch_size=num_samples)
+
+    data_list = []
+    for sample_idx in range(num_samples):
 
-        data_loader = get_dataloaders(batch_size=batch_size, dataset=dataset_name, shuffle=shuffle)
+        if num_samples == 1:
+            return data_loader.dataset[samples_to_retrieve]
```
Comment (linesd) [inline main_viz.py]: Should this not be before the loop? Not a big deal but then it's not evaluated every iteration.

File: viz/visualize.py [hunk 2]
```diff
@@ -8,15 +8,17 @@
 
 from utils.datasets import get_background
 from viz.latent_traversals import LatentTraverser
-from viz.viz_helpers import (reorder_img, read_avg_kl_from_file, add_labels,
+from viz.viz_helpers import (reorder_img, read_loss_from_file, add_labels,
                              upsample, make_grid_img)
-
+import pdb
```
Comment (linesd) [inline viz/visualize.py]: To be removed

File: viz/visualize.py [hunk 3]
```diff
@@ -33,15 +35,22 @@ def __init__(self, model, dataset, model_dir=None, save_images=True):
 
         dataset : str
             Name of the dataset.
+
+        loss_of_interest : str
+            The loss type (as saved in the log file) to order the latent dimensions by and display (optionally)
+
+        display_loss_per_dim : bool
```
Comment (linesd) [inline viz/visualize.py]: This should be is_display_loss_per_dim under the Yann convention for good code writing.

File: viz/visualize.py [hunk 4]
```diff
@@ -33,15 +35,22 @@ def __init__(self, model, dataset, model_dir=None, save_images=True):
...
         self.model = model
         self.device = next(self.model.parameters()).device
         self.latent_traverser = LatentTraverser(self.model.latent_dim)
         self.save_images = save_images
         self.model_dir = model_dir
         self.dataset = dataset
+        self.loss_of_interest = loss_of_interest
 
-    def show_disentanglement_fig2(self, reconstruction_data, latent_sweep_data, heat_map_data, latent_order=None, heat_map_size=(32, 32), filename='imgs/show_disentanglement.png', size=8, sample_latent_space=None):
+    def show_disentanglement_fig2(self, reconstruction_data, latent_sweep_data, heat_map_data, latent_order=None, heat_map_size=(32, 32), filename='show_disentanglement.png', size=8, sample_latent_space=None):
```
Comment (linesd) [inline viz/visualize.py]: Please write this def over multiple lines so it can all be seen.

File: viz/visualize.py [hunk 5]
```diff
@@ -216,6 +226,15 @@ def recon_and_traverse_all(self, data, filename='imgs/recon_and_traverse.png'):
         data : torch.Tensor
             Data to be reconstructed. Shape (N, C, H, W)
 
+        num_increments : int
+            Number of incremental steps to take in the traversal
+
+        reorder_latent_dims : bool
+            If the latent dimensions should be reordered or not
```
Comment (linesd) [inline viz/visualize.py]: Same thing here for the bool naming convention.

File: viz/visualize.py [hunk 6]
```diff
@@ -216,6 +226,15 @@ def recon_and_traverse_all(self, data, filename='imgs/recon_and_traverse.png'):
...
+
+        display_loss_per_dim : bool
```
Comment (linesd) [inline viz/visualize.py]: Same thing here for the bool naming convention.

File: viz/viz_helpers.py [hunk 7]
```diff
@@ -45,53 +46,88 @@ def reorder_img(orig_img, reorder, by_row=True, img_size=(3, 32, 32), padding=2)
...
-                  fill=(0, 0, 0))
-    return traversal_images_with_text
+                      int(((latent_idx+2) / (len(sorted_list)+2) + \
+                            1 / (2 * (len(sorted_list)+2))) * all_traversal_im.height)),
+                    text="{}={}".format(label_name, latent_dim),
+                    fill=(0,0,0),
+                    font=fnt)
+    return traversal_images_with_text   
+
+
```
Comment (linesd) [inline viz/viz_helpers.py]: Remove these blank lines

**Comment by linesd** [comment]
Did a quick review and happy for this to be merged once comments addressed.

File: viz/visualize.py (hunk 4 above)
Comment (alecokas) [reply viz/visualize.py]: I'll do this but I am expecting Bart to open a PR on this very soon

**Pull request merged**